src/util/timelock.rs

```diff
@@ -0,0 +1,264 @@
+///
+/// This is nLockTime as defined in BIP-65.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Absolute {
+    /// A block height absolute timelock.
+    Block {
+        height: u32,
+    },
+    /// A UNIX timestamp absolute timelock.
+    Time {
+        seconds: u32,
+    },
+}
+
+impl Absolute {
+    /// Constructs an absolute timestamp from an nLockTime value.
+    pub fn from_n_lock_time(n: u32) -> Self {
+        if n < LOCKTIME_THRESHOLD {
+            Absolute::Block { height: n }
+        } else {
+            Absolute::Time { seconds: n }
+        }
+    }
+
+    /// Returns true if both timelocks are the same unit i.e., both block based or both time based.
+    pub fn is_same_unit(&self, other: Self) -> bool {
+        use self::Absolute::*;
+
+        match (self, other) {
+            (Block {..}, Block {..} ) | (Time {..}, Time {..}) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if this timelock is locked at the specific time/height `age`.
+    ///
+    /// # Errors
+    ///
+    /// If this timelock and `age` are different units i.e., one represents a
+    /// block height and the other represents a block time.
+    pub fn is_locked(&self, age: u32) -> Result<bool, Error> {
+        self.is_expired(age).map(|b| !b)
+    }
+
+    /// Returns true if this timelock is expired at the specific time/height `age`.
+    ///
+    /// # Errors
+    ///
+    /// If this timelock and `age` are different units i.e., one represents a
+    /// block height and the other represents a block time.
+    pub fn is_expired(&self, age: u32) -> Result<bool, Error> {
+        use self::Absolute::*;
+
+        let age = Absolute::from(age);
+
+        match (*self, age) {
+            (Block { height: ref self_height }, Block { height: age_height }) => {
+                Ok(age_height >= *self_height)
+            }
+            (Time { seconds: ref self_seconds }, Time { seconds: age_seconds }) => {
+                Ok(age_seconds >= *self_seconds)
+            }
+            _ => Err(Error::InvalidComparison),
+        }
+    }
+
+    /// Returns the original nLockTime value.
+    pub fn to_u32(&self) -> u32 {
+        match *self {
+            Absolute::Block { ref height } => *height,
+            Absolute::Time { ref seconds } => *seconds,
+        }
+    }
+}
+
+impl From<u32> for Absolute {
+    /// Creates an absolute timelock from nLockTime.
+    fn from(n: u32) -> Self {
+        Self::from_n_lock_time(n)
+    }
+}
+
+impl fmt::Display for Absolute {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Absolute::Block { ref height } => write!(f, "{}", height),
+            Absolute::Time { ref seconds } => write!(f, "{}", seconds),
+        }
+    }
+}
+
+/// Relative time lock, either after n blocks or after duration (time) as defined by BIP-112.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Rel(u32);
```

Comment:
Why not call it `Relative`? Doesn't seem excessively long and is more readable.

Reply:
Yep, we are going with long names `Absolute` and `Relative`.